library/std/src/net/ip.rs

```diff
@@ -116,10 +115,10 @@ pub struct Ipv4Addr {
 /// assert_eq!("::1".parse(), Ok(localhost));
 /// assert_eq!(localhost.is_loopback(), true);
 /// ```
-#[derive(Copy)]
+#[derive(Copy, Clone, PartialEq, Eq, Hash)]
 #[stable(feature = "rust1", since = "1.0.0")]
 pub struct Ipv6Addr {
-    inner: c::in6_addr,
+    octets: [u8; 16],
```

Comment:
Why not use a u128 here?

Reply:
Any benefit of doing that? It looks like the value is accessed as an array more often, so this felt like a more natural representation. Also it lowers the memory alignment from 4 to 1.

Reply:
It is not entirely clear to me that a lowered alignment is actually beneficial. At least given the fact there will now be a fair number of conversions between this type and the libc one when interacting with the system APIs, the alignment of 1 effectively forces (especially on architectures with strict alignment requirements) a `memcpy` for each such conversion.

Reply:
If you are talking about performance I would argue it's not a valid argument unless proven. As everything regarding optimization. Can you find any usage of this type where it's converted to/from the system representation where said conversion takes up more than a tiny fraction of the time the entire syscall takes?

I'm not saying my benchmarks are perfect. But I have at least tried to check this, and I have not found anything indicating this PR makes anything slower: https://internals.rust-lang.org/t/why-are-socketaddrv4-socketaddrv6-based-on-low-level-sockaddr-in-6/13321/13

I'm not saying alignment 1 is strictly better than 4. Just that I can't find any downsides. And it would be artificial to try to keep it at 4 unless needed, since the most natural representation does not have alignment 4 automatically.

Reply:
If the representation was a `u128` then checking if an address is in a prefix like `::ffff/96` (IPv4-mapped addresses) could be efficiently computed with:

```rust
let mask: u128 = u128::MAX << (128 - 96);
self.inner & mask == Ipv6Addr::from("::ffff").inner
```

Still possible if the representation is `[u8; 16]` using `u128::from_be_bytes`, but because of the alignment that might require a copy.

Reply:
The soon to be released Go 1.18 includes a whole new IP package called `net/netip`: https://github.com/golang/go/blob/master/src/net/netip/netip.go

A blogpost on its design can be found here (the package was originally called `inet.af/netaddr` here, because it was not in the stdlib): https://tailscale.com/blog/netaddr-new-ip-type-for-go/

Importantly for this discussion is that using a pair of uint64 (Go does not have native uint128 support) was quite a bit faster than using an array of bytes. See commit message here for details: https://github.com/inetaf/netaddr/commit/4eb479db13f8b816537f38c664776b193c7a86ec 

I think it's definitely worth running similar benchmarks here.

Reply:
I agree that we should run benchmarks and try to find the best internal representation. But I don't think it should block this PR. Unless we think the PR as is will mean performance regressions I think it brings so many other nice things to the table that it can be ignored for now and iterated upon later. The memory layout of these types are not part of the public API and can change any number of times after this PR is merged.

Reply:
Regarding perf regressions, this is most likely used around syscalls and thus irrelevant. Various Rust APIs like `File` already **heap allocate** on syscalls based on same arguments and heap allocations are significantly slower than a few `mov`s here.